ngcc: new cluster worker entry-point

petebacondarwin opened the pull request:
This refactoring simplifies the clustering by providing a new entry-point file for cluster workers.

This reduces the complexity of the code (not having to check for `cluster.isMaster` so much) and also removes the need to do parse and execute as much code in the worker processes.

petebacondarwin committed:
refactor(ngcc): move command line option parsing to its own file

petebacondarwin committed:
refactor(ngcc): move analyze and compile functions into their own files

petebacondarwin committed:
refactor(ngcc): move pathMapping processing to utils

petebacondarwin committed:
refactor(ngcc): create new entry-point for cluster workers

petebacondarwin committed:
Revert "fix(ngcc): do not spawn unlocker processes on cluster workers (#36569)"

This reverts commit 66effde9f3f9796f9507cd80a85692f0e04e96f5.

petebacondarwin committed:
refactor(ngcc): simplify cluster PackageJsonUpdater

petebacondarwin committed:
fix(ngcc): display unlocker process output in sync mode

The change in e041ac6f0d8c519d104601b2b49c008721c074b7
to support sending unlocker process output to the main ngcc
console output prevents messages require that the main process
relinquishes the event-loop to allow the `stdout.on()` handler to
run.  This results in none of the messages being written when ngcc
is run in `--no-async` mode, and some messages failing to be
written if the main process is killed (e.g. ctrl-C).

It appears that the problem with Windows and detached processes
is known - see https://github.com/nodejs/node/issues/3596#issuecomment-250890218.
But in the meantime, this commit is a workaround, where non-Windows
`inherit` the main process `stdout` while on Windows it reverts
to the async handler approach, which is better than nothing.

gkalpak commented on packages/compiler-cli/ngcc/src/execution/cluster/master.ts:
```diff
@@ -10,7 +10,7 @@
 import * as cluster from 'cluster';
 
-import {resolve} from '../../../../src/ngtsc/file_system';
+import {FileSystem} from '../../../../src/ngtsc/file_system';
```
OOC, why this change?

petebacondarwin replied:
The primary reason was that I started using `resolve()` in the constructor of the `ClusterMaster` and tests started failing because the tests had not called `setFileSystem()`. While I could fix the tests, it made me realise that if this class is relying upon `FileSystem` being setup we ought to make that dependency explicit.

gkalpak commented on packages/compiler-cli/ngcc/src/locking/lock_file_with_child_process/index.ts:
```diff
@@ -82,14 +82,15 @@ export class LockFileWithChildProcess implements LockFile {
-    unlocker.stderr.on('data', data => process.stderr.write(data));
-
+    const isWindows = process.platform === 'win32';
+    const unlocker = fork(
+                         this.fs.resolve(__dirname, './unlocker.js'), [path, logLevel],
+                         {detached: true, stdio: isWindows ? 'pipe' : 'inherit'}) as
+        ChildProcessByStdio<Writable, Readable, Readable>;
+    if (isWindows) {
+      unlocker.stdout.on('data', process.stdout.write.bind(process.stdout));
+      unlocker.stderr.on('data', process.stderr.write.bind(process.stderr));
+    }
```
The `ChildProcessByStdio<Writable, Readable, Readable>` type is not correct any more. (It only has stderr/stdout on Windows.)

You could check for `stderr`/`stdout` explicitly:

```suggestion
    const unlocker = fork(
                         this.fs.resolve(__dirname, './unlocker.js'), [path, logLevel],
                         {detached: true, stdio: isWindows ? 'pipe' : 'inherit'});
    if (unlocker.stdout !== null && unlocker.stderr !== null) {
      unlocker.stdout.on('data', process.stdout.write.bind(process.stdout));
      unlocker.stderr.on('data', process.stderr.write.bind(process.stderr));
    }
```

Or even:

```ts
unlocker.stdout?.on('data', process.stdout.write.bind(process.stdout));
unlocker.stderr?.on('data', process.stderr.write.bind(process.stderr));
```

petebacondarwin replied:
Really like the second option! Thanks.

gkalpak commented on packages/compiler-cli/ngcc/src/main.ts:
```diff
@@ -105,11 +104,7 @@ export function mainNgcc({
-  // NOTE: To avoid file corruption, ensure that each `ngcc` invocation only creates _one_ instance
-  //       of `PackageJsonUpdater` that actually writes to disk (across all processes).
-  //       This is hard to enforce automatically, when running on multiple processes, so needs to be
-  //       enforced manually.
```
I still think this comment is valuable.

petebacondarwin replied:
Yes but it doesn't feel right to have it there... since the PackageJsonUpdater` that is used in `main` will always write to disk... 

I will put a version of this comment in `worker.ts` instead. OK?

gkalpak commented on packages/compiler-cli/ngcc/test/execution/cluster/package_json_updater_spec.ts:
```diff
@@ -23,204 +23,175 @@ runInEachFileSystem(() => {
+    describe('constructor()', () => {
+      it('should throw an error if used on a cluster master', () => {
+        runAsClusterMaster(true);
+        expect(() => new ClusterWorkerPackageJsonUpdater()).toThrowError('');
```
Why use an empty string? We should either use the actual error or not pass an argument at all.

petebacondarwin replied:
Oops!

gkalpak commented on packages/compiler-cli/ngcc/src/execution/cluster/worker.ts:
```diff
@@ -5,58 +5,102 @@
-
 /// <reference types="node" />
 
+import {readConfiguration} from '@angular/compiler-cli/src/perform_compile';
```
Shouldn't this be a relative import?

petebacondarwin replied:
Yep!

gkalpak commented on packages/compiler-cli/ngcc/test/execution/cluster/worker_spec.ts:
```diff
@@ -35,131 +35,116 @@ describe('ClusterWorker', () => {
-      });
+    it('should throw an error', async () => {
+      await expectAsync(runWorker(mockLogger, createCompileFnSpy))
+          .toBeRejectedWithError('Tried to run cluster worker on the master process.');
```
Jasmine 3.5 #ftw :tada:

petebacondarwin replied:
💪

gkalpak commented on packages/compiler-cli/ngcc/src/execution/cluster/worker.ts:
```diff
@@ -5,58 +5,102 @@
+    logger = new ConsoleLogger(LogLevel.info),
+    pathMappings,
+    errorOnFailedEntryPoint = false,
+    enableI18nLegacyMessageIdFormat = true,
```
I would rather bake the defaults into `parseCommandLineOptions()`. Otherwise, we may end up with different values on the master vs workers with unexpected behavior (in the future).

WDYT?

petebacondarwin replied:
The trouble with that is that the sync form of the CLI calls `mainNgcc()` directly and so we need to keep that API stable.

petebacondarwin replied:
I moved all the default calculations into the `setupShared()` function so that should keep main and workers aligned.

gkalpak commented on packages/compiler-cli/ngcc/src/execution/cluster/worker.ts:
```diff
@@ -5,58 +5,102 @@
+          tsConfigPath !== null ? readConfiguration(tsConfigPath || projectPath) : null;
+
+      if (pathMappings === undefined) {
+        pathMappings = getPathMappingsFromTsConfig(tsConfig, projectPath);
```
I am little worried about having identical logic in two places. Everything in this `try` block up to this point is also done in `main.ts`. Instead of having to keep the logic consistent between two places, I would rather have it in one place and share it between master/workers.

petebacondarwin replied:
Yeah, this bothered me too. I will see if I can make it more DRY.

petebacondarwin replied:
Done.

gkalpak reviewed (approved):
Yay! 🎊 👌 💯

petebacondarwin committed:
fixup! fix(ngcc): display unlocker process output in sync mode

petebacondarwin committed:
fixup! refactor(ngcc): simplify cluster PackageJsonUpdater

petebacondarwin committed:
fixup! refactor(ngcc): create new entry-point for cluster workers

petebacondarwin committed:
refactor(ngcc): moved shared setup into a single function

The `main.ts` and `worker.ts` had duplicate logic, which has now been
moved to a single function called `getSharedSetup()`.

petebacondarwin committed:
fixup! refactor(ngcc): create new entry-point for cluster workers
